src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -843,7 +843,7 @@ IsLocallyAccessibleCitusLocalTable(Oid relationId)
 		return false;
 	}
 
-	List *shardIntervalList = LoadShardIntervalList(relationId);
```

Comment:
It seems not all of the callers are locking the shard list. Are we sure it is ok to start locking everywhere?

Reply:
I quitted replacing `LoadShardIntervalListWithRetry` for every `LoadShardIntervalList` as it is too complex to handle deadlocks for each caller.

Instead, I tried to replan non-fastpath queries given inside a distributed plan. I call `planner` to get rerouted plan after acquiring locks. (I could achieve that for Modify non-fastpaths but had deadlock for Select non-fastpaths)